src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,196 @@
+                // Find the buffer we're copying from.
+                T[] buffer = i < 0 ? _first : i < _buffers.Count ? _buffers[i] : _current;
+                
+                // Copy until we satisfy count, or we reach the end of the buffer.
+                int toCopy = Math.Min(count, buffer.Length);
+                Array.Copy(buffer, 0, array, arrayIndex, toCopy);
+
+                // Increment variables to that position.
+                count -= toCopy;
+                arrayIndex += toCopy;
+            }
+        }
+
+        /// <summary>
+        /// Adds an item to this builder.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>
+        /// Use <see cref="Add"/> if adding to the builder is a bottleneck for your use case.
+        /// Otherwise, use <see cref="SlowAdd"/>.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void SlowAdd(T item) => Add(item);
+
+        /// <summary>
+        /// Returns an array representation of this builder.
+        /// </summary>
+        public T[] ToArray()
+        {
+            if (_count == _first.Length)
+            {
+                // No resizing to do.
+                return _first;
+            }
+
+            var array = new T[_count];
+            CopyTo(array, 0, _count);
+            return array;
+        }
+        
+        private T[] AllocateBuffer()
+        {
+            // - On the first few adds, simply resize _first.
+            // - When we pass ResizeLimit, allocate ResizeLimit elements for _current
+            //   and start reading into _current. Set _index to 0.
+            // - When _current runs out of space, add it to _buffers and repeat the
+            //   above step, except with _current.Length * 2.
+
+            Debug.Assert(_index == _current.Length, $"{nameof(AllocateBuffer)} was called, but there's more space.");
+
+            if (_count < ResizeLimit)
+            {
+                // We haven't passed ResizeLimit. Resize _first, copying over the previous items.
+                Debug.Assert(_current == _first && _count == _first.Length);
+
+                int nextCapacity = _count == 0 ? StartingCapacity : _count * 2;
+
+                _current = new T[nextCapacity];
+                Array.Copy(_first, 0, _current, 0, _count);
+                _first = _current;
+            }
+            else
+            {
+                Debug.Assert(_count == ResizeLimit ^ _current != _first);
+
+                int nextCapacity;
+                if (_count == ResizeLimit)
+                {
+                    nextCapacity = ResizeLimit;
+                }
+                else
+                {
+                    _buffers.Add(_current);
+                    nextCapacity = _current.Length * 2;
+                }
+
+                _current = new T[nextCapacity];
+                _index = 0;
+            }
+            
+            return _current;
```

Comment:
Why have AllocateBuffer return _current?  We don't always use the result... why not just have the call sites that need _current access _current?